modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/colocated/GridDhtColocatedLockFuture.java

```diff
@@ -1702,94 +1702,117 @@ void onResult(GridNearLockResponse res) {
             if (res.clientRemapVersion() != null) {
                 assert cctx.kernalContext().clientNode();
 
-                IgniteInternalFuture<?> affFut =
-                    cctx.shared().exchange().affinityReadyFuture(res.clientRemapVersion());
+                if (res.compatibleRemapVersion()) {
+                    if (tx != null) {
+                        // Versions are compatible.
+                        cctx.shared().exchange().
+                            lastAffinityChangedTopologyVersion(res.clientRemapVersion(), tx.topologyVersionSnapshot());
 
-                cctx.time().waitAsync(affFut, tx == null ? 0 : tx.remainingTime(), (e, timedOut) -> {
-                    if (errorOrTimeoutOnTopologyVersion(e, timedOut))
-                        return;
+                        tx.onRemap(res.clientRemapVersion(), false);
 
-                    try {
-                        remap();
-                    }
-                    finally {
-                        cctx.shared().txContextReset();
-                    }
-                });
-            }
-            else {
-                int i = 0;
+                        // Use remapped version for all subsequent mappings.
+                        synchronized (this) {
```

Comment:
It seems that `mappings` should be guarded by `synchronized (GridDhtColocatedLockFuture.this)`